pkg/plugins/repo/client.go

```diff
@@ -26,15 +26,15 @@ type Client struct {
 	log log.PrettyLogger
 }
 
-func newClient(skipTLSVerify bool, logger log.PrettyLogger) *Client {
+func NewClient(skipTLSVerify bool, logger log.PrettyLogger) *Client {
 	return &Client{
 		httpClient:          makeHttpClient(skipTLSVerify, 10*time.Second),
 		httpClientNoTimeout: makeHttpClient(skipTLSVerify, 0),
 		log:                 logger,
 	}
 }
 
-func (c *Client) download(_ context.Context, pluginZipURL, checksum string, compatOpts CompatOpts) (*PluginArchive, error) {
+func (c *Client) Download(_ context.Context, pluginZipURL, checksum string, compatOpts ...CompatOpts) (*PluginArchive, error) {
```

Comment:
It's a bit unclear to me why this function as well as the other functions that accept `compatOpts` have been turned into variadic functions. It looks like those functions only take the first element or ignore it at all.

Was it so the `CompatOpts` is "optional", or am I missing something?

Reply:
Exactly. I wish there was a better optional pattern, but in some cases we don't need to send any options through as we don't need to worry about compatibility. Otherwise, it makes it pretty confusing when consuming this API otherwise

Reply:
This could be misleading, can we at least loop over the `compatOpts` if there are more than one? (or return an error, in case someone uses more than one)

Reply:
Ok cool. I personally find a bit confusing pattern of using variadic functions for optional args, I would prefer something like:

```go
func (c *Client) download(_ context.Context, pluginZipURL, checksum string, compatOpts *CompatOpts) (*PluginArchive, error) {
// ...
}

func (c *Client) Download(ctx context.Context, pluginZipURL, checksum string, compatOpts CompatOpts) (*PluginArchive, error) {
    return c.download(ctx, pluginZipURL, &compatOpts)
}

func (c *Client) DownloadWithoutCompatibility(ctx context.Context, pluginZipURL, checksum string) (*PluginArchive, error) {
    return c.download(ctx, pluginZipURL, nil)
}
```

but I think it's okay :+1:

Reply:
Cool let me play with it and see what fits best 👍

Reply:
Okay I did something and I'm not 100% on it. Please both take a look and let me know what you think 😄

Reply:
what exactly was the change? Code now looks good to me but maybe I overlooked something

Reply:
Basically instead of variadic args, now the API will always expect a `CompatOpts`. But I changed how CompatOpts are constructed by un-exporting the fields and creating two options. OS and Arch are always mandatory, and Grafana version is technically optional (but Grafana server will always need to set it). I can add some better validation logic to capture this with the Grafana plugin installer component.

```go 
func NewCompatOpts(grafanaVersion, os, arch string) CompatOpts {
	return CompatOpts{grafanaVersion: grafanaVersion, arch: arch, os: os}
}

func NewSystemCompatOpts(os, arch string) CompatOpts {
	return CompatOpts{arch: arch, os: os}
}
```

I might tidy this further to be something like:

```go
func NewSystemCompatOpts(os, arch string) CompatOpts {
	return CompatOpts{arch: arch, os: os}
}

func (o CompatOpts) WithGrafanaVersion(grafanaVersion string) CompatOpts
    o.grafanaVersion = grafanaVersion
    return o
}
```